[quant] Rewrite of the scalar mul

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24873 [quant] Rewrite of the scalar mul**

Pull Request resolved: https://github.com/pytorch/pytorch/pull/24873

Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

Commit:
Rewrite of the scalar mul

zafartahirov commented:
This version is not tested yet -- battery dying :)

Commit:
Update on "Rewrite of the scalar mul"


Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

jerryzh168 commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -98,11 +90,10 @@ class QMulScalar final : public c10::OperatorKernel {
                     double scale, int64_t zero_point) {
     TORCH_CHECK(qa.qscheme() == kPerTensorAffine ||
-              qa.qscheme() == kPerTensorSymmetric,
+                qa.qscheme() == kPerTensorSymmetric,
```
I'm going to remove kPerTensorSymmetric in backend, so you can remove this check

Commit:
Update on "[quant] Rewrite of the scalar mul"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24873

Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

llyfacebook commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -49,24 +49,16 @@ Tensor _mul_out(Tensor& out, const Tensor& self, const Tensor& other) {
-    });
-  });
-  return out;
+Tensor& _mul_scalar_(Tensor& self, Scalar other) {
```
Also add the ReluFused logic back?

Commit:
Update on "[quant] Rewrite of the scalar mul"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24873

Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

Commit:
Update on "[quant] Rewrite of the scalar mul"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24873

Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

Commit:
Update on "[quant] Rewrite of the scalar mul"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24873

Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

llyfacebook commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -49,24 +49,20 @@ Tensor _mul_out(Tensor& out, const Tensor& self, const Tensor& other) {
+  float other_float;
+  other_float = other.toFloat();
+  if (self.qscheme() == kPerTensorAffine) {
+    double new_scale = self.q_scale() / other_float;
```
Should this be multiply instead of divide?

lly-zero-one reviewed (changes requested):
the new scale should be the old_scale multiplied with float. By the way, can we do the numeric and performance benchmark for each added op?

Commit:
Update on "[quant] Rewrite of the scalar mul"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24873

Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

Commit:
Update on "[quant] Rewrite of the scalar mul"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24873

Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

Commit:
Update on "[quant] Rewrite of the scalar mul"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24873

Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

Commit:
Update on "[quant] Rewrite of the scalar mul"


Pull Request resolved: https://github.com/pytorch/pytorch/pull/24873

Differential Revision: [D16906728](https://our.internmc.facebook.com/intern/diff/D16906728)

zafartahirov commented:
Works now

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -267,17 +268,49 @@ def test_qadd_relu_different_qparams(self):
+        C_relu_ref = torch.relu(C_relu_ref)
+        C_relu_hat = mul_scalar_relu(qA, b)
+
+        self.assertEqual(C_ref, C_hat.dequantize(), prec=1,
```
prec seems too loose. Do we need 1?
Can we match the reference implementation to get a better match?

zafartahirov replied:
I think at this point this is OK. Once we have a more rigorous guidelines for reference implementation and once we have a clear understanding how to mimic the double rounding, we can change that. Right now, it works with `prec=0.1`, but I think this is data specific

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -49,24 +49,33 @@ Tensor _mul_out(Tensor& out, const Tensor& self, const Tensor& other) {
+    return self;
+  }
+
+  TORCH_CHECK(other_float > 0,
```
Should we support this first prior to landing?

zafartahirov replied:
I would rather land that, and let someone else take over from this point. Otherwise, this PR will be stuck

raghuramank100 reviewed (changes requested):
Lets not rush this one, will ask for others to take this up.

zafartahirov commented:
> Lets not rush this one, will ask for others to take this up.

OK, in that case someone needs to take over it

zafartahirov commented:
Closing as outdated in favor of #26937
